### ISSUE TITLE
kubectl apply for service errors on spec.ports[0].name

### ISSUE THREAD

**Issue by geriBatai** [kubernetes/kubernetes#28225]
Kubectl fails modifying existing service resource when changing it's ports. To reproduce:

kubectl version 1.2.4
api version 1.2.2

original spec, for kubectl create:

```
apiVersion: v1
kind: Service
metadata:
  name: myservice
  labels:
    name: myservice
spec:
  ports:
  - port: 80
    protocol: TCP
    name: tcp-80
    targetPort: 80
  selector:
    name: myservice
```

modified spec, for kubectl apply:
```
apiVersion: v1
kind: Service
metadata:
  name: myservice
  labels:
    name: myservice
spec:
  ports:
  - port: 81
    protocol: TCP
    name: tcp-81
    targetPort: 81
  selector:
    name: myservice
```

Running kubectl apply on the second definition generates the following error:

```
spec.ports[0].name: Required value
```

Debug info shows that it fails PATCH command, because generated diff omits "name". Excerpt from request body:

```
"spec":{"ports":[{"$patch":"delete","port":81},{"name":"tcp-port","port":80,"protocol":"TCP
","targetPort":80}]
```

**Comment by lavalamp** [comment]
@foxish this might be a decent starter bug.

**Comment by AdoHe** [comment]
@geriBatai I think you should use `kubectl apply` to create the original resource, please give a try.

**Comment by geriBatai** [comment]
@AdoHe I'm not really following. There's a bug somewhere in apply diff code (my naive understanding on the issue says that "name" is dropped somewhere because it's optional for a single port) so it wouldn't matter whether you apply or create the first object.

You can always work around issues, e.g. use `kubectl patch`, but that's not what this ticket is about.

**Comment by foxish** [comment]
@lavalamp, I cannot seem to reproduce this with newer releases.

**Comment by lavalamp** [comment]
Can't reproduce because it makes a different patch now?

**Comment by foxish** [comment]
I intercepted the request at the apiserver. I tried using client version 1.2.4 as indicated.
The PATCH looks like:

```
{"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{"kind":"Service","apiVersion":"v1","metadata":{"name":"myservice","creationTimestamp":null,"labels":{"name":"myservice"}},"spec":{"ports":[{"name":"tcp-81","protocol":"TCP","port":81,"targetPort":81}],"selector":{"name":"myservice"}},"status":{"loadBalancer":{}}}"},"creationTimestamp":null},"spec":{"ports":[{"name":"tcp-81","port":81,"protocol":"TCP","targetPort":81}]}}
```
It contains spec.ports[0].name as can be seen, and throws no errors.
It is the same with client version 1.2.2 as well.

**Comment by geriBatai** [comment]
Command execution log attached. True, to get this situation now I had to do

```
$ kubectl create -f s1.yaml
$ kubectl apply -f s2.yaml
$ kubectl apply -f s1.yaml
```

[typescript.txt](https://github.com/kubernetes/kubernetes/files/365825/typescript.txt)

**Comment by foxish** [comment]
It appears this issue was fixed in 1.2.4 by this [PR](https://github.com/kubernetes/kubernetes/commit/70386e4924ab392c2cf9cfb597f2982c39322126).
Adding `patchStrategy:"merge" patchMergeKey:"port"` to the JSON tag fixes the issue. Should we backport this into 1.2.2?

**Comment by lavalamp** [comment]
1.2.4 *is* a patch release of 1.2.2 (we haven't yet needed to do an e.g. 1.2.2.1, and would only consider that for an extreme security problem I think). So the answer is just to update to 1.2.4 or later (1.2.6 was just released). Thanks for investigating!
